File: docs-examples/example-java/src/test/java/io/micronaut/mqtt/docs/custom/annotation/ProductListener.java
package io.micronaut.mqtt.docs.custom.annotation;

// tag::imports[]
import io.micronaut.mqtt.annotation.MqttSubscriber;
import io.micronaut.mqtt.annotation.Topic;

import java.util.*;
// end::imports[]

import io.micronaut.context.annotation.Requires;

@Requires(property = "spec.name", value = "CorrelationSpec")
// tag::clazz[]
@MqttSubscriber
public class ProductListener {

    Set<String> messages = Collections.synchronizedSet(new HashSet<>());

    @Topic("product")
    public void receive(byte[] data, @Correlation byte[] correlation) { // <1>
        messages.add(new String(correlation));
    }
}
// end::clazz[]

File: docs-examples/example-kotlin/src/test/kotlin/io/micronaut/mqtt/docs/custom/annotation/ProductListener.kt
package io.micronaut.mqtt.docs.custom.annotation

// tag::imports[]
import io.micronaut.mqtt.annotation.MqttSubscriber
import io.micronaut.context.annotation.Requires
import io.micronaut.mqtt.annotation.Topic
import java.nio.charset.StandardCharsets

import java.util.Collections
import java.util.HashSet
// end::imports[]

@Requires(property = "spec.name", value = "CorrelationSpec")
// tag::clazz[]
@MqttSubscriber
class ProductListener {

    val messages: MutableSet<String> = Collections.synchronizedSet(HashSet())

    @Topic("product")
    fun receive(@Correlation data: ByteArray) { // <1>
        messages.add(String(data, StandardCharsets.UTF_8))
    }
}
// end::clazz[]
